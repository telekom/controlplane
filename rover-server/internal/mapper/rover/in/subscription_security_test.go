// Copyright 2025 Deutsche Telekom IT GmbH
//
// SPDX-License-Identifier: Apache-2.0

package in

import (
	"github.com/gkampitakis/go-snaps/snaps"
	. "github.com/onsi/ginkgo/v2"
	. "github.com/onsi/gomega"
	roverv1 "github.com/telekom/controlplane/rover/api/v1"

	"github.com/telekom/controlplane/rover-server/internal/api"
)

var _ = Describe("Subscription Security Mapper", func() {
	Context("mapSubscriptionSecurity", func() {
		It("must map BasicAuth security correctly", func() {
			// Given
			input := api.ApiSubscription{
				BasePath: "/test",
			}
			basicAuth := api.BasicAuth{
				Username: "testuser",
				Password: "testpass",
			}
			input.Security = api.Security{}
			err := input.Security.FromBasicAuth(basicAuth)
			Expect(err).To(BeNil())

			output := &roverv1.ApiSubscription{}

			// When
			mapSubscriptionSecurity(input, output)

			// Then
			Expect(output.Security).ToNot(BeNil())
			Expect(output.Security.M2M).ToNot(BeNil())
			Expect(output.Security.M2M.Basic).ToNot(BeNil())
			Expect(output.Security.M2M.Basic.Username).To(Equal("testuser"))
			Expect(output.Security.M2M.Basic.Password).To(Equal("testpass"))
			snaps.MatchSnapshot(GinkgoT(), output.Security)
		})

		It("must map OAuth2 client credentials correctly", func() {
			// Given
			input := api.ApiSubscription{
				BasePath: "/test",
			}
			oauth2 := api.Oauth2{
				ClientId:     "client-id",
				ClientSecret: "client-secret",
				ClientKey:    "client-key",
			}
			input.Security = api.Security{}
			err := input.Security.FromOauth2(oauth2)
			Expect(err).To(BeNil())

			output := &roverv1.ApiSubscription{}

			// When
			mapSubscriptionSecurity(input, output)

			// Then
			Expect(output.Security).ToNot(BeNil())
			Expect(output.Security.M2M).ToNot(BeNil())
			Expect(output.Security.M2M.Client).ToNot(BeNil())
			Expect(output.Security.M2M.Client.ClientId).To(Equal("client-id"))
			Expect(output.Security.M2M.Client.ClientSecret).To(Equal("client-secret"))
			Expect(output.Security.M2M.Client.ClientKey).To(Equal("client-key"))
			snaps.MatchSnapshot(GinkgoT(), output.Security)
		})

		It("must map OAuth2 username/password correctly", func() {
			// Given
			input := api.ApiSubscription{
				BasePath: "/test",
			}
			oauth2 := api.Oauth2{
				Username: "testuser",
				Password: "testpass",
			}
			input.Security = api.Security{}
			err := input.Security.FromOauth2(oauth2)
			Expect(err).To(BeNil())

			output := &roverv1.ApiSubscription{}

			// When
			mapSubscriptionSecurity(input, output)

			// Then
			Expect(output.Security).ToNot(BeNil())
			Expect(output.Security.M2M).ToNot(BeNil())
			Expect(output.Security.M2M.Basic).ToNot(BeNil())
			Expect(output.Security.M2M.Basic.Username).To(Equal("testuser"))
			Expect(output.Security.M2M.Basic.Password).To(Equal("testpass"))
			snaps.MatchSnapshot(GinkgoT(), output.Security)
		})

		It("must map OAuth2 scopes correctly", func() {
			// Given
			input := api.ApiSubscription{
				BasePath: "/test",
			}
			oauth2 := api.Oauth2{
				Scopes: []string{"read", "write"},
			}
			input.Security = api.Security{}
			err := input.Security.FromOauth2(oauth2)
			Expect(err).To(BeNil())

			output := &roverv1.ApiSubscription{}

			// When
			mapSubscriptionSecurity(input, output)

			// Then
			Expect(output.Security).ToNot(BeNil())
			Expect(output.Security.M2M).ToNot(BeNil())
			Expect(output.Security.M2M.Scopes).To(ContainElements("read", "write"))
			snaps.MatchSnapshot(GinkgoT(), output.Security)
		})

		It("must handle invalid security discriminator", func() {
			// Given
			input := api.ApiSubscription{
				BasePath: "/test",
			}
			input.Security = api.Security{} // Invalid security without proper initialization

			output := &roverv1.ApiSubscription{}

			// When
			mapSubscriptionSecurity(input, output)

			// Then
			Expect(output.Security).To(BeNil()) // Should not set security when there's an error
		})

		It("must handle nil security", func() {
			// Given
			input := api.ApiSubscription{
				BasePath: "/test",
			}
			// Security is not set

			output := &roverv1.ApiSubscription{}

			// When
			mapSubscriptionSecurity(input, output)

			// Then
			Expect(output.Security).To(BeNil()) // Should not set security when it's nil
		})

		It("must map combined OAuth2 configuration correctly", func() {
			// Given
			input := api.ApiSubscription{
				BasePath: "/test",
			}
			oauth2 := api.Oauth2{
				ClientId:     "client-id",
				ClientSecret: "client-secret",
				Scopes:       []string{"read", "write"},
			}
			input.Security = api.Security{}
			err := input.Security.FromOauth2(oauth2)
			Expect(err).To(BeNil())

			output := &roverv1.ApiSubscription{}

			// When
			mapSubscriptionSecurity(input, output)

			// Then
			Expect(output.Security).ToNot(BeNil())
			Expect(output.Security.M2M).ToNot(BeNil())
			Expect(output.Security.M2M.Client).ToNot(BeNil())
			Expect(output.Security.M2M.Client.ClientId).To(Equal("client-id"))
			Expect(output.Security.M2M.Scopes).To(ContainElements("read", "write"))
			snaps.MatchSnapshot(GinkgoT(), output.Security)
		})
	})
})
