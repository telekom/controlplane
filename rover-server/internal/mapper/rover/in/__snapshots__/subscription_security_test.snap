
[Subscription Security Mapper mapSubscriptionSecurity must map BasicAuth security correctly - 1]
&v1.SubscriberSecurity{
    M2M: &v1.SubscriberMachine2MachineAuthentication{
        Client: (*v1.OAuth2ClientCredentials)(nil),
        Basic:  &v1.BasicAuthCredentials{Username:"testuser", Password:"testpass"},
        Scopes: nil,
    },
}
---

[Subscription Security Mapper mapSubscriptionSecurity must map OAuth2 client credentials correctly - 1]
&v1.SubscriberSecurity{
    M2M: &v1.SubscriberMachine2MachineAuthentication{
        Client: &v1.OAuth2ClientCredentials{ClientId:"client-id", ClientSecret:"client-secret", ClientKey:"client-key"},
        Basic:  (*v1.BasicAuthCredentials)(nil),
        Scopes: nil,
    },
}
---

[Subscription Security Mapper mapSubscriptionSecurity must map OAuth2 username/password correctly - 1]
&v1.SubscriberSecurity{
    M2M: &v1.SubscriberMachine2MachineAuthentication{
        Client: (*v1.OAuth2ClientCredentials)(nil),
        Basic:  &v1.BasicAuthCredentials{Username:"testuser", Password:"testpass"},
        Scopes: nil,
    },
}
---

[Subscription Security Mapper mapSubscriptionSecurity must map OAuth2 scopes correctly - 1]
&v1.SubscriberSecurity{
    M2M: &v1.SubscriberMachine2MachineAuthentication{
        Client: (*v1.OAuth2ClientCredentials)(nil),
        Basic:  (*v1.BasicAuthCredentials)(nil),
        Scopes: {"read", "write"},
    },
}
---

[Subscription Security Mapper mapSubscriptionSecurity must map combined OAuth2 configuration correctly - 1]
&v1.SubscriberSecurity{
    M2M: &v1.SubscriberMachine2MachineAuthentication{
        Client: &v1.OAuth2ClientCredentials{ClientId:"client-id", ClientSecret:"client-secret", ClientKey:""},
        Basic:  (*v1.BasicAuthCredentials)(nil),
        Scopes: {"read", "write"},
    },
}
---
